tests/test_estimator.py

```diff
@@ -74,29 +93,128 @@ def test_free_energy_estimator():
+            barostat,
             lambda_schedule,
             100,
             100,
-            barostat
+            beta,
+            "test"
         )
 
-        value_and_grad_fn = jax.value_and_grad(estimator.deltaG, argnums=1, has_aux=True)
-        (dG, _), sys_grad = value_and_grad_fn(mdl, sys_params) # run fwd, store result, and run bwd
+        dG, bar_dG_err, results = estimator_abfe.deltaG(mdl, sys_params)
 
-        grad_fn = jax.grad(estimator.deltaG, argnums=1, has_aux=True)
-        grad, _ = grad_fn(mdl, sys_params)
+        return dG**2
+
+
+    for client in [None, CUDAPoolClient(1)]:
+        
+        value_and_grad_fn = jax.value_and_grad(loss_fn)
+        dG, sys_grad = value_and_grad_fn(sys_params) # run fwd, store result, and run bwd
+
+        grad_fn = jax.grad(loss_fn)
+        grad = grad_fn(sys_params)
 
         assert len(grad) == 2
         assert grad[0].shape == sys_params[0].shape
+        assert grad[1].shape == sys_params[1].shape
+
+
+def test_free_energy_estimator_with_endpoint_correction():
+
+    n_atoms = 15
+    x0 = np.random.rand(n_atoms, 3)
+    v0 = np.zeros_like(x0)
+
+    n_bonds = 3
+    n_angles = 4
+    n_restraints = 5
+
+    hb_pot, hb_params = get_harmonic_bond(n_atoms, n_bonds)
+    ha_pot, ha_params = get_harmonic_angle(n_atoms, n_angles)
+    rs_pot, rs_params = get_harmonic_restraints(n_atoms, n_restraints)
+
+    sys_params = [hb_params, ha_params, rs_params]
+    unbound_potentials = [hb_pot, ha_pot, rs_pot]
+
+    masses = np.random.rand(n_atoms)
+
+    box = np.eye(3, dtype=np.float64)
+
+    seed = 2021
+
+    group_idxs = get_group_indices(get_bond_list(hb_pot))
+
+    temperature = 300.0
+    pressure = 1.0
+
+    integrator = LangevinIntegrator(
+        temperature,
+        1.5e-3,
+        1.0,
+        masses,
+        seed
+    )
+
+    barostat = MonteCarloBarostat(
+        x0.shape[0],
+        pressure,
+        temperature,
+        group_idxs,
+        25,
+        seed
+    )
+
+    beta = 0.125
+
+    lambda_schedule = np.linspace(0, 1.0, 4)
+
+    def loss_fn(sys_params):
+
+        endpoint_correct = True
+        mdl = estimator_abfe.FreeEnergyModel(
+            unbound_potentials,
+            endpoint_correct,
+            client,
+            box,
+            x0,
+            v0,
+            integrator,
+            barostat,
+            lambda_schedule,
+            100,
+            100,
+            beta,
+            "test"
+        )
+
+        dG, bar_dG_err, results = estimator_abfe.deltaG(mdl, sys_params)
+
+        return dG**2
+
+
+    for client in [None, CUDAPoolClient(1)]:
+        
+        value_and_grad_fn = jax.value_and_grad(loss_fn)
+        dG, sys_grad = value_and_grad_fn(sys_params) # run fwd, store result, and run bwd
+
+        grad_fn = jax.grad(loss_fn)
+        grad = grad_fn(sys_params)
+
+        assert len(grad) == 3
+        assert grad[0].shape == sys_params[0].shape
         assert grad[1].shape == sys_params[1].shape
```

Comment:
Comment on expectation for `grad[2].shape, sys_params[2].shape`?

Reply:
addressed in `9336d6c`